common/preconditioner/isai_kernels.hpp.inc

```diff
@@ -0,0 +1,220 @@
+    for (int i = 0; i < subwarp_size; ++i) {
+        trisystem[trisystem_start + subwarp_id + i * subwarp_size] =
+            zero<ValueType>();
+    }
+    this_block.sync();
+
+    if (row >= num_rows) {
+        return;
+    }
+
+    const auto i_row_begin = i_row_ptrs[row];
+    const auto i_row_end = i_row_ptrs[row + 1] - i_row_begin < subwarp_size
+                               ? i_row_ptrs[row + 1]
+                               : i_row_begin + subwarp_size;
+    const auto num_row_elems = i_row_end - i_row_begin;
+
+    if (subwarp_id < num_row_elems) {
+        const auto col = i_col_idxs[i_row_begin + subwarp_id];
+        const auto m_row_end = m_row_ptrs[col + 1];
+        auto m_row_ptr = m_row_ptrs[col];
+        auto i_row_ptr = i_row_begin;
+        int idx{trisystem_start +
+                subwarp_id};  // write to trisystem in row major
+
+        while (m_row_ptr < m_row_end && i_row_ptr < i_row_end &&
+               idx < end_trisystem) {
+            const auto sparsity_col = i_col_idxs[i_row_ptr];
+            const auto m_col = m_col_idxs[m_row_ptr];
+            if (sparsity_col == m_col) {
+                trisystem[idx] = m_values[m_row_ptr];
+                idx += subwarp_size;
+                ++m_row_ptr;
+                ++i_row_ptr;
+            } else if (m_col < sparsity_col) {
+                ++m_row_ptr;
+            } else {
+                ++i_row_ptr;
+                idx += subwarp_size;
+            }
+        }
+    }
+
+    this_block.sync();
+
+    // Now, read a full col of `trisystem` into local registers, which will be
+    // row elements after this (implicit) transpose
+    ValueType local_row[subwarp_size];
+    const auto trisystem_col_start =
+        trisystem_start + subwarp_id * subwarp_size;
+    for (int i = 0; i < subwarp_size; ++i) {
+        const auto intermediate_idx = subwarp_id + i;
+        const auto coalesced_idx = intermediate_idx % subwarp_size;
+        local_row[coalesced_idx] =
+            trisystem[trisystem_col_start + coalesced_idx];
+    }
+
+    const auto rhs = trs_solve(num_row_elems, local_row, subwarp);
+
+    // Write back:
+    if (subwarp_id < num_row_elems) {
+        const auto idx = i_row_begin + subwarp_id;
+        if (gko::isfinite(rhs)) {
+            i_values[idx] = rhs;
+        } else {
+            i_values[idx] =
+                i_col_idxs[idx] == row ? one<ValueType>() : zero<ValueType>();
+        }
+    }
+}
+
+
+template <int subwarp_size, int subwarps_per_block, typename ValueType,
+          typename IndexType>
+__global__
+__launch_bounds__(subwarp_size *subwarps_per_block) void generate_l_inverse(
+    IndexType num_rows, const IndexType *__restrict__ m_row_ptrs,
+    const IndexType *__restrict__ m_col_idxs,
+    const ValueType *__restrict__ m_values, IndexType *__restrict__ i_row_ptrs,
+    IndexType *__restrict__ i_col_idxs, ValueType *__restrict__ i_values)
+{
+    auto trs_solve = [](IndexType num_elems,
```

Comment:
In the same vein as the Jacobi kernels: How about making this whole solve fixed-sized (num_elems = subwarp_size) and filling the rest of the data with zeros and diagonal ones? For not too imbalanced matrices, this could drastically improve code generation and allow for the loops to be completely unrolled (even though I am not sure how well the optimizer could actually do, since we have a rather long critical path here)

Maybe we shouldn't do that in this PR, but improve it later?

Reply:
I tried that, however, the register usage just went up slightly, while the local memory usage stayed the same (I actually did not look at the PTX output, just the ressource usage). I don't think it was properly unrolled, maybe the program would have been too large that the compiler chose against it.
I also did not spend too much time trying to optimize it perfectly since this is only the generation, the actual application is very fast since it will just be an SpMV.